text/0000-prefetch-hook.md

```diff
@@ -0,0 +1,79 @@
+- Start Date: 2015-10-01
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+The waterfall way in which Ember resolves a transition's model hooks can result in unnecessary blocking.
+Performance wins can be achieved by allowing routes to decouple initiating network requests from waiting for their completion.
+
+# Motivation
+
+Currently, the `model` hook is Ember's way for routes to request data.
+However, a route's `model` hook is not entered until it's parent route has resolved.
+This is suboptimal for nested application structures.
+Child routes are forced to wait for their parents to receive data before they may make requests for their own data.
+This occurs even when the child doesn't make use of its parents' data.
+
+This RFC proposes adding a new `Route#prefetch` hook, which is used to allow routes to make requests in parallel.
+Child routes that make use of `prefetch` will resolve faster since their data can resolve at the same time as (or before) their parents'.
+The semantics and ordering of `Route`'s existing model hooks (`beforeModel`, `model`, `afterModel`) are preserved.
+
+See this demo for a visualization: http://nickiaconis.github.io/ember-parallel-model-demo/
+
+# Detailed design
+
+A `prefetch` hook is added to `Route`.
+It takes the same parameters as the `model` hook.
+Like the `model` hook, it is not called if an object is passed to the transition.
+
+```javascript
+App.PostRoute = Ember.Route.extend({
+  prefetch(params) {
+    return Ember.$.get(`/api/posts/${params.id}`);
+  }
+});
+
+App.PostCommentsRoute = Ember.Route.extend({
+  prefetch(params, transition) {
+    return Ember.$.get(`/api/posts/${transition.params.post.id}/comments`);
+  }
+});
+```
+
+The default functionality of the `model` hook is modified to return the prefetched data if it exists.
```

Comment:
I'm beginning to come around again on the idea of `this.prefetched('otherRoute')`. Previously, all examples I had seen or devised had involved a route that was a direct descendent of the parent whose data it needed, in which case using the `model` hook is mostly synonymous with using the `prefetch` hook and awaiting a call to `prefetched`. In the case when the child route is the second descendent (or further removed), it makes sense to be able to do this in the `prefetch` hook.

I worry that a `Route#prefetched` method will be (ab)used to get values that are already found in the parameters of parent routes. The correct way of doing that is to use `Route#paramsFor`, but I'm not sure it's ubiquitously known as evidenced by this [SO question](http://stackoverflow.com/questions/22376994/ember-deeply-nested-routes-do-not-keep-parent-dynamic-parameter/27871764), two of the answers for which suggest using the parent route's model. And then there's [this tutorial](https://www.airpair.com/javascript/emberjs-using-ember-cli#6-direct-access-to-router-states) that instead uses the private API of the transition.

However, someone who would to use the `prefetched` method to get a parameter value would probably also use the `model` hook with `modelFor` instead of the `prefetch` hook if the `prefetched` method is not provided.